File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/graphics/UISurface.java
package com.hello.uxml.tools.framework.graphics;

import com.hello.uxml.tools.framework.BlendMode;
import com.hello.uxml.tools.framework.BorderRadius;
import com.hello.uxml.tools.framework.Rectangle;

/**
 * Represents a basic 2D compositing element for rendering and UI input.
 *
 * @author ferhat
 */
public interface UISurface {

  /**
   * Sets the target to be called for input events on the surface.
   */
  void setTarget(UISurfaceTarget target);

  /**
   * Clears the render list.
   */
  void clearRenderList();

  /**
   * Add rectangle to render list.
   */
  void drawRect(Brush brush, Pen pen, Rectangle rect);

  /**
   * Add rounded rectangle to render list.
   */
  void drawRect(Brush brush, Pen pen, Rectangle rect, BorderRadius borderRadius);

  /**
   * Add ellipse to render list.
   */
  void drawEllipse(Brush brush, Pen pen, Rectangle rect);

  /**
   * Add path to render list.
   */
  void drawPath(Brush brush, Pen pen, VecPath path);

  /**
   * Add line to render list.
   */
  void drawLine(Pen pen, double xFrom, double yFrom, double xTo, double yTo);

  /**
   * Add child surface.
   */
  void addChild(UISurface surface);

  /**
   * Remove child surface.
   */
  void removeChild(UISurface surface);

  /**
   * Changes child's zorder.
   */
  void setChildDepth(UISurface child, int depth);

  /**
   * Returns number of children.
   */
  int getChildCount();

  /**
   * Returns child at index.
   */
  UISurface getChild(int index);

  /**
   * Returns parent surface.
   */
  UISurface getParentSurface();

  /**
   * Sets parent surface.
   */
  void setParentSurface(UISurface parentSurface);

  /**
   * Removes item from display list.
   */
  void close();

  /**
   * Creates a child UISurface
   */
  UISurface createChildSurface();

  /**
   * Creates a child UITextSurface
   */
  UISurface createChildTextSurface();

  /**
   * Creates a child UIImageSurface
   */
  UISurface createChildImageSurface();

  /**
   * Gets or sets surface location.
   */
  void setLayout(Rectangle layoutRect);

  /**
   * Sets surface visibility.
   */
  void setVisible(boolean visible);

  /**
   * Refreshes view.
   */
  void updateView();

  /**
   * Refreshes view area.
   */
  void updateView(double x, double y, double width, double height);

  /**
   * Sets opacity of surface 0(transparent)..1(opaque)
   */
  void setOpacity(double opacity);

  /**
   * Sets blend mode of surface.
   */
  void setBlendMode(BlendMode blendMode);

  /**
   * Adds a filter to surface.
   */
  void addFilter(Filter filter);
  /**
   * Removes a filter from surface.
   */
  void removeFilter(Filter filter);

  /**
   * Performs hit testing on self and child elements. If successful,
   * return true and fills the hitResult object.
   */
  boolean hitTest(double x, double y, HitTestResult hitResult);

  /**
   * Sets hit testing strategy.
   * <p>HITTEST_CONTENT = 0 , checks for visible items
   * <p>HITTEST_BOUNDS = 1, checks against surface bounds. Typically used for
   * invisible areas that are buttons.
   */
  public void setHitTestMode(int mode);
}

File: framework/lib/ui_surface.dart
part of uxml;

/**
 * Represents a basic 2D surface for rendering and UI input.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
abstract class UISurface {
  /**
   * Sets the target to be called for input events on the surface.
   */
  set target(UISurfaceTarget target);
  UIElement get target;

  /** Adds a child surface. Returns child. */
  UISurface addChild(UISurface child);
  UISurface insertChild(int index, UISurface child);
  bool removeChild(UISurface child);
  void reparentChild(UISurface child, [int index = -1]);

  // Called by child when a host element is created for surface.
  void _hostElementCreated(UISurface surface);
  void _onInitSurface();

  int get childCount;
  UISurface childAt(int index);

  // Sets the location and size of surface relative to parent.
  void setLocation(int targetX, int targetY, int targetWidth, int targetHeight);

  void setBackground(Brush brush);
  void setBorderRadius(BorderRadius border);
  set clipChildren(bool clip);
  /** Delays updates to add multiple children efficiently. */
  void lockUpdates(bool lock);

  int layoutX;
  int layoutY;
  int layoutW;
  int layoutH;
  // Tag is a value attached to the surface to enable platform
  // specific debugging.
  set tag(String value);
  UITransform renderTransform;
  UITransform get surfaceTransform;
  set surfaceTransform(UITransform surfaceTransform);

  /**
   * Clears graphics surface.
   */
  void clear();

  /**
   * Drawing API.
   */
  void drawPath(VecPath path, Brush brush, Pen pen, Margin nineSlice);
  void drawRect(num x, num y, num width, num height,
      Brush brush, Pen pen, BorderRadius borderRadius);
  void drawEllipse(num x, num y, num width, num height, Brush brush, Pen pen);
  void drawLine(Brush brush, Pen pen, num xFrom,num yFrom, num xTo, num yTo);
  void drawBitmap(Object image, num x, num y, num width, num height, bool tile);
  void maskBitmap(Object image, num x, num y, num width, num height, VecPath mask);
  void drawBitmapTransformColor(Object image, num x, num y, num width,
      num height, bool tile, ColorTransform transform);
  void drawImage(String source);

  set hitTestMode(int mode);
  UISurface hitTest(num mouseX, num mouseY);

  UISurface parentSurface;
  Application root;
  bool mouseEnabled;
  bool get enableHitTesting;
  set enableHitTesting(bool);
  bool get enableChildHitTesting;
  set enableChildHitTesting(bool);
  set opacity(num value);
  set visible(bool value);
  bool get visible;
  bool get hasTransform;
  void applyFilters(Filters filters);
  bool setupTopLevelCursor(Cursor cursor);
  void cursorChanged(Cursor cursor);
  void close();
}

abstract class UITextSurface extends UISurface {
  set text(String value);
  set htmlText(String value);
  set wordWrap(bool value);
  set textAlign(int value);
  set visible(bool value);
  set enableHitTesting(bool value);
  set fontName(String value);
  set fontSize(num value);
  set fontBold(bool value);
  set textColor(Color value);
  set onTextLink(EventHandler handler);
  void measureText(String text, num availWidth, num availHeight,
      bool sizeToBold);
  set selectable(bool value);
  void updateTextView();
  int measuredWidth = 0;
  int measuredHeight = 0;
}

abstract class UIEditSurface extends UISurface {
  set text(String value);
  set htmlText(String value);
  set multiline(bool value);
  set wordWrap(bool value);
  set visible(bool value);
  set enableHitTesting(bool value);
  void enableMouseEvents(bool val);
  set fontName(String value);
  set fontSize(num value);
  set fontBold(bool value);
  set textColor(Color value);
  set onTextLink(EventHandler handler);
  void measureText(String text, num availWidth, num availHeight);
  set promptMessage(String prompt);
  set maxChars(int length);
  void initFocus(bool selectAll);
  void focusChanged(bool isFocused);
  int measuredWidth;
  int measuredHeight;
}
